File: Cloud Storage Samples/CloudFiles/java/cloudfiles.java
/*
 * Cloud Storage 2022 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Storage in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudstorage
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import cloudstorage.*;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Scanner;

public class cloudfiles {

    public cloudfiles() {
        Scanner scanner = new Scanner(System.in);
        Cloudfiles cloudfiles = new Cloudfiles();
        Oauth oauth = new Oauth();

        try {
            cloudfiles.addCloudfilesEventListener(new CloudfilesEvents());

            System.out.println("Which cloud service provider would you like to use to send your message?");
            System.out.println("   [0] - Amazon S3");
            System.out.println("   [1] - GoogleStorage");
            System.out.println("   [2] - Azure Blob");

            int servicePrompt = Integer.parseInt(scanner.nextLine());

            // Prompt for authentication information.
            if (servicePrompt == 0) {
                cloudfiles.setServiceProvider(0);

                System.out.print("Enter your Amazon access key:  ");
                cloudfiles.getAccount().setAccessKey(scanner.nextLine());

                System.out.print("Enter your Amazon secret key:  ");
                cloudfiles.getAccount().setSecretKey(scanner.nextLine());
            } else if (servicePrompt == 1) {
                cloudfiles.setServiceProvider(9);
                oauth.setServerAuthURL("https://accounts.google.com/o/oauth2/auth");
                oauth.setServerTokenURL("https://accounts.google.com/o/oauth2/token");

                System.out.print("Enter your OAuth client ID:  ");
                oauth.setClientId(scanner.nextLine());

                System.out.print("Enter your OAuth client secret:  ");
                oauth.setClientSecret(scanner.nextLine());

                System.out.print("Authenticating... ");
                //cloudfiles.Account.Authorization = await oauth.GetAuthorization();
                cloudfiles.getAccount().setAuthMechanism(5);
                cloudfiles.setAuthorization(oauth.getAuthorization());
                System.out.print(" done.");
            } else if (servicePrompt == 2) {
                cloudfiles.setServiceProvider(6);
                oauth.setServerAuthURL("https://login.microsoftonline.com/common/oauth2/v2.0/authorize");
                oauth.setServerTokenURL("https://login.microsoftonline.com/common/oauth2/v2.0/token");

                System.out.print("Enter your OAuth client ID:  ");
                oauth.setClientId(scanner.nextLine());

                System.out.print("Enter your OAuth client secret:  ");
                oauth.setClientSecret(scanner.nextLine());

                System.out.print("Authenticating... ");
                cloudfiles.getAccount().setAuthMechanism(5);
                cloudfiles.setAuthorization(oauth.getAuthorization());
                System.out.print(" done.");
            } else {
                throw new Exception("\nInvalid cloud service provider.\n");
            }

            // Prompt for message information.
            System.out.println("What would you like to do?\n ");
            System.out.println("   [0] - Upload a file");
            System.out.println("   [1] - Download a file");
            System.out.println("   [2] - Delete a file");
            System.out.println("   [3] - Create Directory");
            System.out.println("   [4] - Delete Directory");
            System.out.println("   [5] - Quit");
            servicePrompt = Integer.parseInt(scanner.nextLine());

            String filename;
            String localFile;
            while (servicePrompt != 5) {
                switch (servicePrompt) {
                    case 0:
                        System.out.print("Remote file path: ");
                        filename = scanner.nextLine();
                        System.out.print("Local file path: ");
                        localFile = scanner.nextLine();
                        cloudfiles.setLocalFile(localFile);
                        cloudfiles.setRemoteFile(filename);
                        cloudfiles.upload();
                        System.out.println("Done.");
                        break;
                    case 1:
                        System.out.print("Remote file path: ");
                        filename = scanner.nextLine();
                        System.out.print("Local file path: ");
                        localFile = scanner.nextLine();
                        cloudfiles.setRemoteFile(filename);
                        cloudfiles.setLocalFile(localFile);
                        cloudfiles.download();
                        System.out.println("Done.");
                        break;
                    case 2:
                        System.out.print("File path: ");
                        filename = scanner.nextLine();
                        cloudfiles.deleteFile(filename);
                        System.out.println("Done.");
                        break;
                    case 3:
                        System.out.print("Directory name: ");
                        filename = scanner.next();
                        cloudfiles.makeDirectory(filename);
                        System.out.println("Done.");
                        break;
                    case 4:
                        System.out.print("Directory name: ");
                        filename = scanner.nextLine();
                        cloudfiles.removeDirectory(filename);
                        System.out.println("Done.");
                        break;
                    default:
                        throw new IllegalStateException("Unexpected value: " + servicePrompt);
                }
                System.out.println("What would you like to do?\n ");
                System.out.println("   [0] - Upload a file");
                System.out.println("   [1] - Download a file");
                System.out.println("   [2] - Delete a file");
                System.out.println("   [3] - Create Directory");
                System.out.println("   [4] - Delete Directory");
                System.out.println("   [5] - Quit");
                servicePrompt = Integer.parseInt(scanner.nextLine());
            }
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }

        System.out.println("\nPress any key to exit...");

    }
    public static void main(String[] args) {
        try {
            new cloudfiles();
        } catch (Exception e) {
            System.out.println("Exception: " + e.getMessage());
        }
    }


}

 class CloudfilesEvents extends DefaultCloudfilesEventListener {

    Scanner scn = new Scanner(System.in);

    public void SSLServerAuthentication(CloudfilesSSLServerAuthenticationEvent e) {
        if (e.accept) return;
        System.out.print("Server provided the following certificate:\nIssuer: " + e.certIssuer + "\nSubject: " + e.certSubject + "\n");
        System.out.print("The following problems have been determined for this certificate: " + e.status + "\n");
        System.out.print("Would you like to continue anyways? [y/n] ");
        if (scn.nextLine().charAt(0) == 'y') e.accept = true;
    }
}

class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }

  static String prompt(String label, String punctuation, String defaultVal)
  {
	System.out.print(label + " [" + defaultVal + "] " + punctuation + " ");
	String response = input();
	if(response.equals(""))
		return defaultVal;
	else
		return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof CloudStorageException) {
      System.out.print(" (" + ((CloudStorageException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }
}

File: Cloud Storage Samples/CloudFiles/cpp/cloudfiles.cpp
/*
 * Cloud Storage 2022 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Storage in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudstorage
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <iostream>
#include <fstream>
#include <string>
#include "../../include/cloudstorage.h"

using namespace std;

#define LINE_LEN 80
#define MESSAGE_LEN 1024


int main()
{
    char command[LINE_LEN];     // user's command

    CloudFiles cloudfiles1;
    OAuth oauth1;

    // Welcome Message
    printf("Welcome to the Cloud Files demo.\n");
    printf("------------------------------------------------------------\n");

    printf("Which cloud service provider would you like to use to send your message? \n  [0] - Amazon S3\n  [1] - Google Storage\n  [2] - Azure Blob\n");
    fgets(command, LINE_LEN, stdin);
    if (std::stoi(command) == 0) {
        cloudfiles1.SetServiceProvider(0);
        printf("Enter your Amazon Access key: ");
        fgets(command, LINE_LEN, stdin);
        cloudfiles1.SetAccountAccessKey(command);
        printf("Enter your Amazon secret key: ");
        cloudfiles1.SetAccountSecretKey(command);
        fgets(command, LINE_LEN, stdin);
    }
    else if (std::stoi(command) == 1) {
        cloudfiles1.SetServiceProvider(9);
        oauth1.SetServerAuthURL("https://accounts.google.com/o/oauth2/auth");
        oauth1.SetServerTokenURL("https://accounts.google.com/o/oauth2/token");
        printf("Enter your OAuth Client ID: ");
        fgets(command, LINE_LEN, stdin);
        oauth1.SetClientId(command);
        printf("Enter your OAuth client secret: ");
        fgets(command, LINE_LEN, stdin);
        oauth1.SetClientSecret(command);
        printf("Authenticating...\n");
        cloudfiles1.SetAuthorization(oauth1.GetAuthorization());
        printf("Authentication successful.\n");
    }
    else if (std::stoi(command) == 2) {
        cloudfiles1.SetServiceProvider(6);
        oauth1.SetServerAuthURL("https://login.microsoftonline.com/common/oauth2/v2.0/authorize");
        oauth1.SetServerTokenURL("https://login.microsoftonline.com/common/oauth2/v2.0/token");
        printf("Enter your OAuth Client ID: ");
        fgets(command, LINE_LEN, stdin);
        oauth1.SetClientId(command);
        printf("Enter your OAuth client secret: ");
        fgets(command, LINE_LEN, stdin);
        oauth1.SetClientSecret(command);
        printf("Authenticating...\n");
        cloudfiles1.SetAuthorization(oauth1.GetAuthorization());
        printf("Authentication successful.\n");
    }
    else {
        throw std::invalid_argument("Invalid mail service provider.");
    }

    printf("What would you like to do? \n  [0] - Upload a file\n  [1] - Download a file\n  [2] - Delete a file\n  [3] - Create Directory\n  [4] - Delete Directory\n  [5] - Quit");

    fgets(command, LINE_LEN, stdin);

    while (std::stoi(command) != 5) {
        if (std::stoi(command) == 0) {
            printf("Local file path: ");
            fgets(command, LINE_LEN, stdin);
            cloudfiles1.SetLocalFile(command);
            printf("Remote file path: ");
            fgets(command, LINE_LEN, stdin);
            cloudfiles1.SetRemoteFile(command);
            cloudfiles1.Upload();
            printf("Done.\n");
        }
        else if (std::stoi(command) == 1) {
            printf("Local file path: ");
            fgets(command, LINE_LEN, stdin);
            cloudfiles1.SetLocalFile(command);
            printf("Remote file path: ");
            fgets(command, LINE_LEN, stdin);
            cloudfiles1.SetRemoteFile(command);
            cloudfiles1.Download();
            printf("Done.\n");
        }
        else if (std::stoi(command) == 2) {
            printf("File path: ");
            fgets(command, LINE_LEN, stdin);
            cloudfiles1.DeleteFile(command);
            printf("Done.\n");
        }
        else if (std::stoi(command) == 3) {
            printf("Directory name: ");
            fgets(command, LINE_LEN, stdin);
            cloudfiles1.MakeDirectory(command);
            printf("Done.\n");
        }
        else if (std::stoi(command) == 4) {
            printf("Directory name: ");
            fgets(command, LINE_LEN, stdin);
            cloudfiles1.RemoveDirectory(command);
            printf("Done.\n");
        }
        else if (std::stoi(command) == 5) {
            break;
        }
        else {
            printf("Invalid command");
        }
        printf("What would you like to do? \n  [0] - Upload a file\n  [1] - Download a file\n  [2] - Delete a file\n  [3] - Create Directory\n  [4] - Delete Directory\n  [5] - Quit");

        fgets(command, LINE_LEN, stdin);
    }
    exit(0);
}





